File: my-folder/3659-count-paths-with-the-given-xor-value/solution.py
class Solution:
    def countPathsWithXorValue(self, grid: List[List[int]], k: int) -> int:
        n, m = len(grid), len(grid[0])
        MOD = 10**9 + 7
        
        dp = [[[-1]*16 for _ in range(m)] for _ in range(n)]
        
        def solve(i, j, path_xor):
            path_xor ^= grid[i][j]
            
            if i == n-1 and j == m-1:
                return 1 if path_xor == k else 0
            
            if dp[i][j][path_xor] != -1:
                return dp[i][j][path_xor]
            
            res = 0
            if i + 1 < n:
                res += solve(i+1, j, path_xor)
            if j + 1 < m:
                res += solve(i, j+1, path_xor)
            
            dp[i][j][path_xor] = res % MOD
            return dp[i][j][path_xor]
        
        return solve(0, 0, 0)

File: my-folder/3659-count-paths-with-the-given-xor-value/solution.cpp
class Solution {
public:
    int modulo = 1e9 + 7;
    int f(vector<vector<int>>& grid, int i, int j, int zor, int k, vector<vector<vector<int>>>& dp)
    {
        if (i < 0 || j<0)
            return 0;
        if (i == 0 && j == 0)
            if ((zor ^ grid[i][j]) == k)
                return 1;
            else
                return 0;
            
        if (dp[i][j][zor] != -1)
            return dp[i][j][zor];

        int xornow = ((zor)^(grid[i][j]));
        int top = f(grid, i - 1, j, xornow, k, dp);
        int right = f(grid, i, j - 1, xornow, k, dp);
        return dp[i][j][zor] = (top + right) % modulo;
 
    }   

    int countPathsWithXorValue(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<vector<int>>> dp(m, vector<vector<int>>(n, vector<int>(16,-1)));
        return f(grid, m - 1, n - 1, 0, k, dp);
    }
};
